File: other/day23.c
// SPDX-License-Identifier: MIT
// Copyright (c) 2020 Chua Hou

#include <stdio.h>
#include <stdlib.h>

int main(void)
{
	// le_mutable_structure[i] = j means the i-th cup is followed by cup j
	int le_mutable_structure[1000000 + 1];

	FILE *file = fopen("input/day23.txt", "r");
	int first = (int) (fgetc(file) - '0');
	int curr = first;
	char c;
	c = fgetc(file);
	do {
		int next = (int) (c - '0');
		le_mutable_structure[curr] = next;
		curr = next;
	} while ((c = fgetc(file)) != '\n');
	fclose(file);

	for (int next = 10; next <= 1000000; next++) {
		le_mutable_structure[curr] = next;
		curr = next;
	}
	le_mutable_structure[curr] = first;

	curr = first;
	for (int i = 0; i < 10000000; i++) {
		int pick1 = le_mutable_structure[curr];
		int pick2 = le_mutable_structure[pick1];
		int pick3 = le_mutable_structure[pick2];
		le_mutable_structure[curr] = le_mutable_structure[pick3];

		int dest = curr == 1 ? 1000000 : curr - 1;
		while (dest == pick1 || dest == pick2 || dest == pick3) {
			dest = dest == 1 ? 1000000 : dest - 1;
		}
		le_mutable_structure[pick3] = le_mutable_structure[dest];
		le_mutable_structure[dest] = pick1;

		curr = le_mutable_structure[curr];
	}

	unsigned long star1 = le_mutable_structure[1];
	unsigned long star2 = le_mutable_structure[star1];
	printf("%lu\n", star1 * star2);

	return 0;
}

File: src/AOC/Days/Day23.hs
-- SPDX-License-Identifier: MIT
-- Copyright (c) 2020 Chua Hou

module AOC.Days.Day23 (solution) where

import           AOC.Solution

import           Control.Monad               ((<=<))
import           Control.Monad.ST            (ST, runST)
import           Data.List                   (sortOn)
import qualified Data.Vector.Unboxed         as V
import qualified Data.Vector.Unboxed.Mutable as VM

-- 0th element is current cup; i-th element is the cup following the i-th cup
type Game = V.Vector Int

stepGame :: Int -> Int -> Game -> Game
stepGame maxCups iters g = runST (V.thaw g >>= go iters >>= V.freeze)
    where
        go :: Int -> VM.MVector s Int -> ST s (VM.MVector s Int)
        go 0 v = pure v
        go i v = do
            { curr     <- v `VM.read` 0
            ; pick1    <- v `VM.read` curr
            ; pick2    <- v `VM.read` pick1
            ; pick3    <- v `VM.read` pick2
            ; next     <- v `VM.read` pick3
            ; let dest =  checkDest (curr - 1) [pick1, pick2, pick3]
            ; dnext    <- v `VM.read` dest
            ; VM.write v curr  next
            ; VM.write v pick3 dnext
            ; VM.write v dest  pick1
            ; VM.write v 0     next
            ; go (i - 1) v
            }
        checkDest x ps
            | x <= 0      = checkDest maxCups ps
            | x `elem` ps = checkDest (x - 1) ps
            | otherwise   = x

solution :: [Int] :=> Maybe Int
solution = simpleSolution
    (Just . map (read . pure) . filter (/= '\n'))
    (showGame . stepGame 9       100      <=< mkGame)
    (getStars . stepGame 1000000 10000000 <=< mkGame . (<> [10..1000000]))
    where
        showGame g = go (g V.!? 1) >>= readMaybe . concatMap show
            where
                go x =   x >>= \x' -> g V.!? x'
                     >>= \case
                         1 -> Just [x']
                         y -> (x':) <$> go (Just y)
        getStars g = do
            { s1 <- g V.!? 1
            ; s2 <- g V.!? s1
            ; return $ s1 * s2
            }
        mkGame (x:xs) = Just $ V.fromList . (x:) . map snd . sortOn fst
                      $ zip (x:xs) (xs <> [x])
        mkGame []     = Nothing
